solr/modules/monitor/src/java/org/apache/solr/monitor/update/MonitorUpdateRequestProcessor.java

```diff
@@ -0,0 +1,257 @@
+              .map(
+                  queryStr ->
+                      new MonitorQuery(
+                          queryId, SimpleQueryParser.parse(queryStr, core), queryStr, Map.of()))
+              .stream()
+              .flatMap(monitorQuery -> decompose(monitorQuery, payload))
+              .map(this::toSolrInputDoc)
+              .collect(Collectors.toList());
+      if (children.isEmpty()) {
+        throw new SolrException(
+            SolrException.ErrorCode.INVALID_STATE, "Query could not be decomposed");
+      }
+      SolrInputDocument firstChild = children.get(0);
+      if (solrInputDocument.hasChildDocuments()) {
+        solrInputDocument.getChildDocuments().clear();
+      }
+      solrInputDocument.addChildDocuments(children.stream().skip(1).collect(Collectors.toList()));
+      if (solrInputDocument.hasChildDocuments()) {
+        solrInputDocument
+            .getChildDocuments()
+            .forEach(
+                child ->
+                    solrInputDocument.forEach(
+                        field -> {
+                          if (!MonitorFields.RESERVED_MONITOR_FIELDS.contains(field.getName())) {
+                            child.addField(field.getName(), field.getValue());
+                          }
+                        }));
+        solrInputDocument
+            .getChildDocuments()
+            .forEach(
+                child ->
+                    child.setField(
+                        indexSchema.getUniqueKeyField().getName(),
+                        child.getFieldValue(MonitorFields.CACHE_ID)));
+      }
+      copyFirstChildToParent(solrInputDocument, firstChild);
+    }
+    super.processAdd(cmd);
+  }
+
+  private void copyFirstChildToParent(SolrInputDocument parent, SolrInputDocument firstChild) {
+    parent.setField(
+        indexSchema.getUniqueKeyField().getName(),
+        firstChild.getFieldValue(MonitorFields.CACHE_ID));
+    for (var firstBornInputField : firstChild) {
+      parent.setField(firstBornInputField.getName(), firstBornInputField.getValue());
+    }
+  }
+
+  private Stream<Document> decompose(MonitorQuery monitorQuery, String payload) {
+    return QCEVisitor.decompose(monitorQuery, MonitorConstants.QUERY_DECOMPOSER).stream()
+        .map(
+            qce -> {
+              Document doc =
+                  presearcher.indexQuery(qce.getMatchQuery(), monitorQuery.getMetadata());
+              doc.add(monitorSchemaFields.getQueryId().createField(qce.getQueryId()));
+              doc.add(monitorSchemaFields.getCacheId().createField(qce.getCacheId()));
+              doc.add(
+                  monitorSchemaFields.getMonitorQuery().createField(monitorQuery.getQueryString()));
+              if (payload != null) {
+                doc.add(monitorSchemaFields.getPayload().createField(payload));
+              }
+              return doc;
+            });
+  }
+
+  private SolrInputDocument toSolrInputDoc(Document subDocument) {
```

Comment:
Got a little stuck from here onwards reading code in this class -- to be continued another time then probably.

Reply:
Ah yes, this is probably the hairiest bit of the integration. I mentioned it briefly in the description but I should have brought more attention to it. I'll admit that I may be relying on some implementation details. Basically, lucene monitor gives you back some fields as token streams. You can "materialize" those back into a solr document field and if you make that field an `IndexableField`, you can pass that `TokenStream` directly to the lucene indexing chain because solr assumes an `IndexableField` needs no adaptation [here](https://github.com/apache/solr/blob/b8364e2309d0fa5120b87b945b70663b4f56ee14/solr/core/src/java/org/apache/solr/update/DocumentBuilder.java#L59-L66). Perhaps this is an implementation detail? Either way, it seemed logical and I thought it would be quite a pain to serialize that token stream returned by lucene-monitor into a more conventional solr document field just to deserialize it again later on.

The motivation behind `DerivedSkipTLogField` is that all lucene-monitor derived fields are just that, derived. So there is no need to write them to the tlog as long as we always perform the derivation. This assumption creates a dependency on the order of the `MonitorUpdateProcessor` which needs to be called out; it needs to be done _after_ the `DistributedUpdateProcessor`. But because derivation is cheap and the alternatives were ugly, I figured this was the best way. I've manually tested shutting down a solr instance with uncommitted changes and then bringing it back up with updates applied successfully. This makes me think this approach will work although automated tests should probably be added if this approach were to be accepted.